polygeist/tools/cgeist/Test/Verification/reinterpret_cast.cpp

```diff
@@ -0,0 +1,66 @@
+#include <cstdint>
+
+// RUN: cgeist --use-opaque-pointers -O0 %s --function=* -S | FileCheck %s
+
+// CHECK-LABEL:   func.func @_Z17reinterpret_floatRf(
+// CHECK-SAME:                                       %[[VAL_0:.*]]: memref<?xf32>) -> memref<?xi32>
+// CHECK:           %[[VAL_1:.*]] = "polygeist.memref2pointer"(%[[VAL_0]]) : (memref<?xf32>) -> !llvm.ptr
+// CHECK:           %[[VAL_2:.*]] = "polygeist.pointer2memref"(%[[VAL_1]]) : (!llvm.ptr) -> memref<?xi32>
+// CHECK:           return %[[VAL_2]] : memref<?xi32>
+// CHECK:         }
+
+int32_t &reinterpret_float(float &f) {
+  return reinterpret_cast<int32_t &>(f);
+}
+
+// CHECK-LABEL:   func.func @_Z17reinterpret_floatf(
+// CHECK-SAME:                                      %[[VAL_0:.*]]: f32) -> i32
+// CHECK:           %[[VAL_1:.*]] = memref.alloca() : memref<1xf32>
+// CHECK:           %[[VAL_2:.*]] = llvm.mlir.undef : f32
+// CHECK:           affine.store %[[VAL_0]], %[[VAL_1]][0] : memref<1xf32>
+// CHECK:           %[[VAL_3:.*]] = "polygeist.memref2pointer"(%[[VAL_1]]) : (memref<1xf32>) -> !llvm.ptr
+// CHECK:           %[[VAL_4:.*]] = llvm.load %[[VAL_3]] : !llvm.ptr -> i32
+// CHECK:           return %[[VAL_4]] : i32
+// CHECK:         }
+
+int32_t reinterpret_float(float f) {
+  return reinterpret_cast<int32_t &>(f);
+}
+
+struct fooint32 { int32_t x; };
+struct foofloat { float x; };
+
+// CHECK-LABEL:   func.func @_Z17reinterpret_floatR8foofloat(
+// CHECK-SAME:                                               %[[VAL_0:.*]]: !llvm.ptr) -> !llvm.ptr
```

Comment:
[note]: I was expecting to see the reference to struct being lowered to a memref<llvm.struct<f32>>, I guess the FE flattens the struct is there is only one member in it?

Reply:
We're generating `llvm.ptr` here as `memref<llvm.struct<...>>` is illegal. I'm assuming we will be using `llvm.ptr` most times, but some parts of the code will use the `memref` construct because they might be creating a `memref` type (not through `getType`) and not checking whether the element type is a `struct`. This is just a guess, though.